### PR TITLE
CT-2258 Offender SARs Record Data Requests

### PR DESCRIPTION
**Pull request by mseedat-moj**
## Description
Allow an Offender SAR to have data requests recorded.
 
## Self-review checklist

* [x] (1) Quick stakeholder demo done OR
* [x] (2) ...bug with before and after screenshots
* [x] (3) Tests passing
* [x] (4) Branch ready to be merged (not work in progress)
* [x] (5) No superfluous changes in diff
* [x] (6) No TODO's without new ticket numbers
 
### Related JIRA tickets
https://dsdmoj.atlassian.net/browse/CT-2258
 
### Manual testing instructions
1. Log in as a Branston User e.g. Brian Rix
2. Create a new Offender SAR 
3. Note 'Record data request' green button in action bar
4. Proceed to add data requests
5. Note case now is in a 'waiting for data' state, and 'Record data request' button is grey colour

### CODE REVIEW

File: app/models/case/sar/offender.rb
Comment (mseedat-moj) [inline app/models/case/sar/offender.rb]: @Polsonby This is the workaround I mentioned regarding preventing GovUKDateFields from throwing the 'no method error valid? on nil class'. I believe it occurs when creating an object using the ActiveRecord relationship. In my case doing `offender_sar.data_requests.create!(...)` (IN A TEST!!) resulted in the `valid? on nil:class` error. This is my proposed work around. I have lost too many hours to this issue.

File: app/models/case/sar/offender.rb
Comment (Polsonby) [inline app/models/case/sar/offender.rb]: Is this easier to read? Seems a little indirect, but not such that I'd complain about it. Ah - just read further down - we use the list again 👍 

But maybe we could have the `acts_as_gov_uk_date(*GOV_UK_DATE_FIELDS)` right after the frozen array so they're next to each other?

File: app/models/case/sar/offender.rb
Comment (Polsonby) [reply app/models/case/sar/offender.rb]: Same here! Looks cool - nice job, I'm sure it will work better than my previous "fix"

File: app/policies/case/sar/offender_policy.rb
Comment (Polsonby) [inline app/policies/case/sar/offender_policy.rb]: Are there ever any cases for `case/sar/offender_policy` that `self.case` could be something other than an offender_sar? Just wondering why not say "true" here

File: spec/controllers/cases/data_requests_controller_spec.rb
Comment (Polsonby) [inline spec/controllers/cases/data_requests_controller_spec.rb]: This looks a wee bit odd - should the user not be a Branston user?

File: spec/services/data_request_service_spec.rb
Comment (Polsonby) [inline spec/services/data_request_service_spec.rb]: :)

File: spec/services/data_request_service_spec.rb
Comment (Polsonby) [inline spec/services/data_request_service_spec.rb]: Something missing here?

**Review by Polsonby — COMMENTED** [review]
Great work, just a few questions

File: app/policies/case/sar/offender_policy.rb
Comment (mseedat-moj) [reply app/policies/case/sar/offender_policy.rb]: errrr..... good point!  I think I did it to explicitly tell the reader you can only record a data request if the current case is an Offender SAR but I suppose that's a given!

File: spec/services/data_request_service_spec.rb
Comment (mseedat-moj) [reply spec/services/data_request_service_spec.rb]: Rectified in #895

File: app/models/case/sar/offender.rb
Comment (mseedat-moj) [reply app/models/case/sar/offender.rb]: I haven't verified that this fix works for all the existing places where the annoying error crops up

File: app/views/cases/data_requests/_data_request.html.slim
Comment (Polsonby) [inline app/views/cases/data_requests/_data_request.html.slim]: Should be h2 for accessibility, I believe - screen reader users expect the headings to be used in sequential order and not miss out levels. But super tiny issue

File: app/views/cases/offender_sar/_data_requests.html.slim
Comment (Polsonby) [inline app/views/cases/offender_sar/_data_requests.html.slim]: I believe this date should be formatted like other dates on the page

```suggestion
                = l data_request.date_requested, format: :default
```

File: app/helpers/cases_helper.rb
Comment (Polsonby) [inline app/helpers/cases_helper.rb]: This still POSTs to the next page - really annoying! But not worth spending time on IMO

**Review by Polsonby — APPROVED** [review]
Great work, Mo 👍 

The headings on the new data request form should be h2 not h3, and the date format on the list of  should be consistent with other dates. Approving anyway but if you're pushing again, perhaps you could address these two tiny points

File: app/helpers/cases_helper.rb
Comment (mseedat-moj) [reply app/helpers/cases_helper.rb]: I will check this before merging to master as I have it definitely being s GET to the next page!

File: app/views/cases/data_requests/_data_request.html.slim
Comment (mseedat-moj) [reply app/views/cases/data_requests/_data_request.html.slim]: This is a partial included on the `cases/show` page which already has `<h2>` elements. Therefore using `<h3>` here.

File: app/helpers/cases_helper.rb
Comment (mseedat-moj) [reply app/helpers/cases_helper.rb]: Can you check this again - I cannot reproduce:
Screen Shot 2019-08-30 at 10 23 38

Comment (Polsonby) [reply app/helpers/cases_helper.rb]: image
It's consistent in Chrome and Safari - click the "Record data request" button and hit reload in the browser (without the web inspector open) and it asks you to resubmit the form - header shows POST not GET. 

But as I said, the code in the view looks fine to me and IMO it's not worth spending time on investigating further just now - just annoying. Presumably there is some UJS overriding the data-method attribute somewhere

File: app/views/cases/data_requests/_data_request.html.slim
Comment (Polsonby) [reply app/views/cases/data_requests/_data_request.html.slim]: That's the other partial - this is the form fields in the new data requests form. When you go to `http://localhost:3000/cases/9/data_requests/new`

File: app/helpers/cases_helper.rb
Comment (mseedat-moj) [reply app/helpers/cases_helper.rb]: Fixed - it's because in the helper I copy pasted a previous link/button, and it has `method: :get` set. This causes some interesting behaviour!

```
    when :record_data_request
      btn_type = @case.current_state == 'data_to_be_requested' ? 'secondary' : 'tertiary'
      link_to 'Record data request',
              new_case_data_request_path(@case),
              id: 'action--record-data-request',
              class: "button-#{btn_type}", method: :get
```

Removing the `method: :get` solves the problem!!

**Comment by mseedat-moj** [comment]
Superceded by Parent PR:

https://github.com/ministryofjustice/correspondence_tool_staff/pull/900

**Pull request closed without merging**
